Notebook decomposition PR, updated

reuben opened the pull request:
First I regenerated the script with jupyter-nbconvert, then I re-applied Tilman's changes to inline and reformat comments, sort imports, things like that. Then I cherry-picked his math support commit to the branch, and finally updated the README.

kdavis-mozilla commented:
Later today, after epoch 20 is finished for the current server run, I'll run a test of this on the server, before I start a deeper review.

kdavis-mozilla commented:
@reuben It's been running on the server for over a day and seems to perform fine. I'll start a deeper review.

kdavis-mozilla commented on README.md:
```diff
@@ -17,7 +16,7 @@ If you have a capable (Nvidia, at least 8GB of VRAM) GPU, it is highly recommend
 ## Training a model
 
-Open a terminal, change to the directory of the DeepSpeech checkout and run `jupyter-notebook DeepSpeech.ipynb`. This should open your default browser with the DeepSpeech notebook. From here, you can alter any variables with regards to what dataset is used, how many training iterations are run and the default values of the network parameters. Then, to train the network, select `Cell` from the not ...
+Open a terminal, change to the directory of the DeepSpeech checkout and run `python DeepSpeech.py`. By default, the code will train on a small sample dataset called LDC93S1, which can be overfitted on a GPU in a few minutes for demonstration purposes.
```
Why was the info

> From here, you can alter any variables with regards to what dataset is used, how many training iterations are run and the default values of the network parameters.

removed?

reuben replied:
No reason, just oversight. I'll add it back.

kdavis-mozilla commented on doc/DeepSpeech.md:
```diff
@@ -0,0 +1,70 @@
+
```
General comment on this approach:

> How will documentation in the `doc` directory stay in sync with the code?

Generally,  if the code and the documentation is in the same place, this process is "easy". However, here the code and documentation are in separate files. This doesn't seem to lend itself to keeping the documentation and code in sync.

Why was something like [Sphinx](http://www.sphinx-doc.org/en/stable/index.html) not used?

reuben replied:
Sphinx looks interesting. I kept @tilmankamp's approach here without really questioning it. Maybe other options are not as easy to publish on GitHub?

kdavis-mozilla replied:
@reuben Maybe that's why @tilmankamp took that approach. I don't know. Also, he's on holiday this week. So we can't ask him.

kdavis-mozilla commented on DeepSpeech.py:
```diff
@@ -0,0 +1,1164 @@
+# ===================
+
+# For an explanation of the meaning of the geometric constants, please refer to
+# doc/Geometry.md
```
Why are the other docs

- `doc/DeepSpeech.md`
-  `doc/ParallelOptimization.md`

Not referenced in this file?

reuben replied:
What do you think about merging doc/DeepSpeech.md into README.md and referencing ParallelOptimization.md from DeepSpeech.py?

kdavis-mozilla replied:
Seems reasonable. But I'd still like the doc and code in one place.

reuben replied:
So I looked into Sphinx + GitHub. It's definitely doable, but it would require an additional build step for *any* documentation changes, not just the math stuff in the doc folder (as they exist in this PR).

IMO, the cleanest option would be setting up a GitHub hook to automatically rebuild the docs on every commit and publish on this repo's gh-pages branch. That way we could keep just updating comments here and docs would magically be available in https://mozilla.github.io/DeepSpeech

What do you think about this approach?

reuben replied:
One alternative is using Sphinx and then simply enabling GitHub's readthedocs service, which automatically updates docs. Downside is that the docs would be published in readthedocs.io, not here in the repo.

reuben commented:
I just pushed a bunch of changes here. Here's a summary:

dc13d4b - Removes gendocs.sh, all the generated SVGs, and renames the `.raw` docs to `.rst`.

0a1d3d4 - Just adds an `if __name__ == "__main__":` block to DeepSpeech.py so it's possible to import it without training the network.

56eee9a - Converts the formatting in the `doc/*.rst` files to be compatible with Sphinx RST. Links, images, code blocks, etc. I also added a backwards version of the RNN diagram because that was bugging me.

d6fb444 - Converts comments before functions to docstrings, where applicable. Since the original comments were written in a literary style, this means the new docs will look a bit weird. But I plan on going over the comments and adjusting the wording later.

af8d3ed - This adds the files needed to generate Sphinx docs. conf.py and index.rst are the important ones, the makefiles are generated.

ee50d82 - The way Sphinx autodoc works requires importing the modules to read their docstrings, so readthedocs.org needs to have our dependencies installed. Luckily, they support specifying a pip requirements file.

ceaccce - Addressing the review comment.

reuben commented:
Here's how the generated docs look like: http://deepspeech.readthedocs.io/en/latest/

kdavis-mozilla commented on doc/DeepSpeech.rst:
```diff
@@ -0,0 +1,76 @@
+trained to ingest speech spectrograms and generate English text transcriptions.
+
+Let a single utterance :math:`x` and label :math:`y` be sampled from a training set
+`S = \{(x^{(1)}, y^{(1)}), (x^{(2)}, y^{(2)}), . . .\}`.
```
Doesn't render correctly. Looks like a Sphinx bug as the LaTex is fine.

reuben replied:
That was my mistake, looks like I missed all the equations that start in the beginning of the line when converting the document. I fixed it.
